numba/core/function_type.py

```diff
@@ -0,0 +1,261 @@
+"""Provides Numba type, FunctionType, that makes functions as
+instances of a first-class function type.
+"""
+
+from numba.extending import typeof_impl
+from numba.extending import models, register_model
+from numba.extending import unbox, NativeValue, box
+from numba.core.imputils import lower_constant, lower_cast
+from numba.core.ccallback import CFunc
+from numba.core import cgutils
+from llvmlite import ir
+from numba.core import types
+from numba.core.types import (FunctionType, UndefinedFunctionType,
+                              FunctionPrototype, WrapperAddressProtocol)
+from numba.core.dispatcher import Dispatcher
+
+
+@typeof_impl.register(WrapperAddressProtocol)
+@typeof_impl.register(CFunc)
+def typeof_function_type(val, c):
+    if isinstance(val, CFunc):
+        sig = val._sig
+    elif isinstance(val, WrapperAddressProtocol):
+        sig = val.signature()
+    else:
+        raise NotImplementedError(
+            f'function type from {type(val).__name__}')
+    return FunctionType(sig)
+
+
+@register_model(FunctionPrototype)
+class FunctionProtoModel(models.PrimitiveModel):
+    """FunctionProtoModel describes the signatures of first-class functions
+    """
+    def __init__(self, dmm, fe_type):
+        if isinstance(fe_type, FunctionType):
+            ftype = fe_type.ftype
+        elif isinstance(fe_type, FunctionPrototype):
+            ftype = fe_type
+        else:
+            raise NotImplementedError((type(fe_type)))
+        retty = dmm.lookup(ftype.rtype).get_value_type()
+        args = [dmm.lookup(t).get_value_type() for t in ftype.atypes]
+        be_type = ir.PointerType(ir.FunctionType(retty, args))
+        super(FunctionProtoModel, self).__init__(dmm, fe_type, be_type)
+
+
+@register_model(FunctionType)
+@register_model(UndefinedFunctionType)
+class FunctionModel(models.StructModel):
+    """FunctionModel holds addresses of function implementations
+    """
+    def __init__(self, dmm, fe_type):
+        members = [
+            # address of cfunc wrapper function:
+            ('addr', types.voidptr),
+            # address of PyObject* referencing the Python function
+            # object:
+            ('pyaddr', types.voidptr),
+        ]
+        super(FunctionModel, self).__init__(dmm, fe_type, members)
+
+
+@lower_constant(types.Dispatcher)
+def lower_constant_dispatcher(context, builder, typ, pyval):
+    return context.add_dynamic_addr(builder, id(pyval),
+                                    info=type(pyval).__name__)
+
+
+@lower_constant(FunctionType)
+def lower_constant_function_type(context, builder, typ, pyval):
+    typ = typ.get_precise()
+
+    if isinstance(pyval, CFunc):
+        addr = pyval._wrapper_address
+        sfunc = cgutils.create_struct_proxy(typ)(context, builder)
+        sfunc.addr = context.add_dynamic_addr(builder, addr,
```

Comment:
The use of `add_dynamic_addr` will disable caching.

Reply:
So it means that if a Numba jit-decorated function uses a first-class function that is passed in via namespace lookup (rather than via argument passing) then such a function cannot be cached.
Is there a way to detect if the user tries to cache such functions so that the exception could be raised at the lower_constant call?

Reply:
> Is there a way to detect if the user tries to cache such functions so that the exception could be raised at the lower_constant call?

Cannot do that consistently. What will happen is that the `add_dynaic_addr` would place something into the LLVM IR such that later the caching layer will raise a warning.

Reply:
OK. I suggest no changes since eventually there will be a warning.